### PR TITLE
chore: use pytest in UT

### PR DESCRIPTION
**Pull request by patriciareinoso**
# Description

Remove `unittest` and use only `pytest` framework to run the unit tests.
There are no modifications in the core content of the tests.
Add gnbsim lib to the coverage path

https://juju.is/docs/sdk/write-a-unit-test-for-a-charm

Actions on this PR:
- Add a `@pytest.fixture` configure harness
- Add a `@pytest.fixture` setup patches and mock harness
- Continue using `Mock`, `call` and `patch` from the `unittest.Mock` library
- Use `assert` instead of `self.assert...` methods
- Add a parameterized test using `@pytest.mark.parametrize`

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

**Review by benhoyt — COMMENTED** [review]
Nice -- this looks very reasonable to me!

One thing I wondered about was the use of `pytest-mock` rather than just keeping on using the stdlib's `Mock` and `@patch` (or you could use pytest's built in `mockeypatch` fixture, but that's a bit limiting). I personally would prefer to avoid the additional 3rd party dependency, but what you have seems fine too.

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -27,560 +26,528 @@ def read_file(path: str) -> str:
...
+    harness.handle_exec("gnbsim", [], result=0)
+    harness.add_storage("config", attach=True)
+    harness.set_can_connect(container="gnbsim", val=True)
+    set_n2_relation_data(harness)
+    harness.evaluate_status()
+    assert harness.charm.unit.status == ActiveStatus()
+
+
+@pytest.fixture
+def harness(mocker):
```
Comment (IronCore864) [inline tests/unit/test_charm.py:67]: This fixture is defined below a few functions that actually use it. The readability can be improved if the fixture is moved to the top, I think.

**Comment by patriciareinoso** [comment]
> Nice -- this looks very reasonable to me!
> 
> One thing I wondered about was the use of `pytest-mock` rather than just keeping on using the stdlib's `Mock` and `@patch` (or you could use pytest's built in `mockeypatch` fixture, but that's a bit limiting). I personally would prefer to avoid the additional 3rd party dependency, but what you have seems fine too.

Initially I added `pytest-mock` because I was having problems with `@patch`. It seems like it is not compatible with `@fixture`
`monkeypatch` was not so easy to implement but I found a way to replace the `@patch` using fixtures. @benhoyt  what do you think ? -> Lines 32-57

**Comment by benhoyt** [comment]
That works!

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -27,560 +29,527 @@ def read_file(path: str) -> str:
...
+    harness.add_storage("config", attach=True)
+    harness.set_can_connect(container="gnbsim", val=True)
+    set_n2_relation_data(harness)
+    harness.evaluate_status()
+    assert harness.charm.unit.status == ActiveStatus()
+
+def create_n2_relation(harness) -> int:
+    """Create a relation between gnbsim and AMF.
 
-class TestCharm(unittest.TestCase):
```
Comment (gruyaume) [inline tests/unit/test_charm.py:31]: I don't know that we should get rid of the fact that all the unit tests for the charm class are in a charm test class. This is quite useful for separating unit tests when we have multiple classes.

For instance, we can go with something like this:

```python

class TestCharm:

    patcher_lightkube_client = patch("lightkube.core.client.GenericSyncClient")
    patcher_k8s_service_patch = patch("charm.KubernetesServicePatch")
    patcher_multus = patch(MULTUS_LIB)
    patcher_gnb_identity = patch(GNB_IDENTITY_LIB)

    @pytest.fixture
    def setup(self) -> Generator[Mock, None, None]:
        self.mock_lightkube_client = TestCharm.patcher_lightkube_client.start()
        self.mock_k8s_service_patch = TestCharm.patcher_k8s_service_patch.start()
        self.mock_multus = TestCharm.patcher_multus.start()
        self.mock_gnb_identity = TestCharm.patcher_gnb_identity.start()
    
    @pytest.fixture()
    def harness(self, setup):
        harness = testing.Harness(GNBSIMOperatorCharm)
        harness.set_model_name(name=NAMESPACE)
        harness.begin()
        yield harness
        harness.cleanup()

    def set_up_active_status_charm(self, harness):
        harness.handle_exec("gnbsim", [], result=0)
        harness.add_storage("config", attach=True)
        harness.set_can_connect(container="gnbsim", val=True)
        self.set_n2_relation_data(harness)
        harness.evaluate_status()
        assert harness.charm.unit.status == ActiveStatus()

    def create_n2_relation(self, harness) -> int:
        """Create a relation between gnbsim and AMF.

        Returns:
            int: Id of the created relation
        """
        amf_relation_id = harness.add_relation(relation_name="fiveg-n2", remote_app="amf")
        harness.add_relation_unit(relation_id=amf_relation_id, remote_unit_name="amf/0")
        return amf_relation_id

    def set_n2_relation_data(self, harness) -> int:
        """Create the N2 relation, set the relation data in the n2 relation and return its ID.

        Returns:
            int: ID of the created relation
        """
        amf_relation_id = self.create_n2_relation(harness)
        harness.update_relation_data(
            relation_id=amf_relation_id,
            app_or_unit="amf",
            key_values={
                "amf_hostname": "amf",
                "amf_port": "38412",
                "amf_ip_address": "1.1.1.1",
            },
        )
        return amf_relation_id

    @pytest.mark.parametrize("config_param", [("usim-opc"),
                                            ("gnb-ip-address"),
                                            ("icmp-packet-destination"),
                                            ("imsi"),
                                            ("mcc"),
                                            ("mnc"),
                                            ("usim-key"),
                                            ("usim-sequence-number"),
                                            ("sd"),
                                            ("tac"),
                                            ("upf-subnet"),
                                            ("upf-gateway")])
    def test_given_invalid_config_when_config_changed_then_status_is_blocked(self, harness, config_param):
        harness.update_config(key_values={config_param: ""})
        harness.evaluate_status()

        assert harness.charm.unit.status == BlockedStatus(
            f"Configurations are invalid: ['{config_param}']"
        )
```

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py [hunk 3]
```diff
@@ -1,30 +1,42 @@
...
 
+class TestGnbIdentityRequires:
 
-class TestGnbIdentityRequires(unittest.TestCase):
+    patcher_gnb_identity = patch(f"{TEST_CHARM_PATH}.fiveg_gnb_identity_available")
+
+    @pytest.fixture()
     def setUp(self) -> None:
+        self.mock_gnb_identity = TestGnbIdentityRequires.patcher_gnb_identity.start()
+        self.mock_gnb_identity.__class__ = BoundEvent
```
Comment (gruyaume) [inline tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py:20]: Why do we need this? ain't there a nicer way?

**Review by gruyaume — APPROVED** [review]
Overall I think this looks quite nice and it is an improvement from what we had. I just have 1 question. I added @ghislainbourgeois to get his opinion and make sure we align before merging and going on to other charms.

**Review by ghislainbourgeois — APPROVED** [review]

File: tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py (hunk 3 above)
Comment (gatici) [reply tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py:20]: framework.observe(someobj.something_happened, self._on_something_happened)
The expected second parameter type is BoundEvent. When we mock it, the following error happens and setting the type to "BoundEvent" is required if we would like to check on emitted events.
```
Framework.observe requires a BoundEvent as second parameter, got <MagicMock name='fiveg_gnb_identity_available' id='140609246438992'>
```

Comment (gatici) [reply tests/unit/lib/charms/sdcore_gnbsim/v0/test_fiveg_gnb_indentity_requirer.py:20]: There tried an alternative approach without setting a type for mocked CharmEvents. In this case, we can not assert on calls for the mocked object but we could check the emitted events by using logs.  

```python
import logging
from unittest.mock import Mock, patch

import pytest
from ops import testing

from lib.charms.sdcore_nrf_k8s.v0.fiveg_nrf import NRFRequirerCharmEvents
from tests.unit.charms.sdcore_nrf.v0.dummy_requirer_charm.src.dummy_requirer_charm import (  # noqa: E501
    DummyFiveGNRFRequirerCharm,
)

DUMMY_REQUIRER_CHARM = "tests.unit.charms.sdcore_nrf.v0.dummy_requirer_charm.src.dummy_requirer_charm.DummyFiveGNRFRequirerCharm"  # noqa: E501
FIVEG_NRF_LIB = "lib.charms.sdcore_nrf_k8s.v0.fiveg_nrf.NRFRequirerCharmEvents"
RELATION_NAME = "fiveg_nrf"
REMOTE_APP_NAME = "dummy-nrf-requirer"
NRF_URL = "https://nrf.example.com"


class TestFiveGNRFRequirer:
    
    @pytest.fixture()
    def setUp(self) -> None:
        self.mock_nrf_requirer_charm_events = NRFRequirerCharmEvents()
        self.mock_nrf_requirer_charm_events.nrf_available = Mock(name='nrf_available')
        self.mock_nrf_requirer_charm_events.nrf_broken = Mock(name='nrf_broken')


    def tearDown(self) -> None:
        patch.stopall()

    @pytest.fixture(autouse=True)
    def harness(self, setUp, request):
        self.harness = testing.Harness(DummyFiveGNRFRequirerCharm)
        self.harness.begin()
        yield self.harness
        self.harness.cleanup()
        request.addfinalizer(self.tearDown)

    def _create_relation(self, remote_app_name: str):
        relation_id = self.harness.add_relation(
            relation_name=RELATION_NAME, remote_app=remote_app_name
        )
        self.harness.add_relation_unit(
            relation_id=relation_id, remote_unit_name=f"{remote_app_name}/0"
        )

        return relation_id

    def test_given_nrf_information_in_relation_data_when_relation_changed_then_nrf_available_event_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)

        relation_data = {
            "url": NRF_URL,
        }
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )

    def test_given_nrf_information_not_in_relation_data_when_relation_changed_then_nrf_available_event_not_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        # self.mock_fiveg_nrf_available.assert_not_called()
        self.mock_nrf_requirer_charm_events.nrf_available.assert_not_called()


    def test_given_invalid_nrf_information_in_relation_data_when_relation_changed_then_nrf_available_event_not_emitted(  # noqa: E501
        self
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        self.mock_nrf_requirer_charm_events.nrf_available.assert_not_called()

    def test_given_invalid_nrf_information_in_relation_data_when_relation_changed_then_error_is_logged(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}
        caplog.set_level(logging.DEBUG)
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        assert "Invalid relation data: {'pizza': 'steak'}" in caplog.messages

    def test_given_nrf_information_in_relation_data_when_get_nrf_url_is_called_then_expected_url_is_returned(  # noqa: E501
        self,
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)

        relation_data = {
            "url": NRF_URL,
        }
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )

        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url == NRF_URL

    def test_given_nrf_information_not_in_relation_data_when_get_nrf_url_then_returns_none(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {}

        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        caplog.set_level(logging.DEBUG)
        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url is None
        assert "Invalid relation data: {}" in caplog.messages

    def test_given_nrf_information_in_relation_data_is_not_valid_when_get_nrf_url_then_returns_none_and_error_is_logged(  # noqa: E501
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        relation_data = {"pizza": "steak"}
        self.harness.update_relation_data(
            relation_id=relation_id, app_or_unit=REMOTE_APP_NAME, key_values=relation_data
        )
        caplog.set_level(logging.DEBUG)
        nrf_url = self.harness.charm.nrf_requirer.nrf_url
        assert nrf_url is None
        assert "Invalid relation data: {'pizza': 'steak'}" in caplog.messages

    def test_given_nrf_relation_created_when_relation_broken_then_nrf_broken_event_emitted(
        self, caplog
    ):
        relation_id = self._create_relation(remote_app_name=REMOTE_APP_NAME)
        caplog.set_level(logging.DEBUG)
        self.harness.remove_relation(relation_id)
        assert "Emitting custom event <NRFBrokenEvent via DummyFiveGNRFRequirerCharm/NRFRequires[fiveg_nrf]/on/nrf_broken[5]>." == caplog.messages[0]

```

**Pull request merged**